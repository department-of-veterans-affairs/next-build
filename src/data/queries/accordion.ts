import { QueryFormatter } from 'next-drupal-query'
import { ParagraphAccordion } from '@/types/drupal/paragraph'
import { Accordion } from '@/types/formatted/accordion'

export const formatter: QueryFormatter<ParagraphAccordion, Accordion> = (
  entity: ParagraphAccordion
) => {
  return {
    id: entity.id,
    header: entity.field_header || null,
    html: entity.field_rich_wysiwyg.processed,
  }
}
