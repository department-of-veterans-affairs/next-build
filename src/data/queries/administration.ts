import { FieldAdministration } from '@/types/drupal/field_type'

export interface Administration {
  entityId: number
  name: string | null
}

/**
 * Formats a Drupal administration field into our standardized Administration interface.
 * 
 * In our Drupal content, the `field_administration` field points to a taxonomy Term
 * that is currently called "Section", but "Section" is only a human label added to
 * it in the Drupal interface, and that name has evolved over time. It was originally,
 * "administration" (and thus the name of the field). Since the only machine name of
 * it is `field_administration`, we'll continue to call it "administration" in our code.
 */
export const formatAdministration = (
  administration: FieldAdministration | null | undefined
): Administration => ({
  entityId: administration?.drupal_internal__tid || null, // "t" is for taxonomy term
  name: administration?.name || null,
})
