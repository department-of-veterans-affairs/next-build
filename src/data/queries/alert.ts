// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter } from 'next-drupal-query'
import { Alert } from '@/types/formatted/alert'

export const formatter: QueryFormatter<BlockAlert, Alert> = (
  entity: BlockAlert
) => {
  return {
    alertType:
      entity.field_alert_type === 'information'
        ? 'info'
        : entity.field_alert_type,
    id: entity.id,
    title: entity.field_alert_title,
    content: {
      header: entity.field_alert_content.field_text_expander,
      text: entity.field_alert_content.field_wysiwyg.processed,
    },
  }
}
