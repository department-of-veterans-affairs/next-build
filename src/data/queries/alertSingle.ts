import { QueryFormatter } from 'next-drupal-query'
import { ParagraphAlertSingle } from '@/types/drupal/paragraph'
import {
  AlertSingle,
  AlertBlock,
  AlertNonReusable,
} from '@/types/formatted/alert'
import { formatParagraph } from '@/lib/drupal/paragraphs'
import { queries } from '.'

// TODO:
//
// EDIT:
// It seems *maybe* that next-drupal-query is designed to handle this already.
// You can pass an `id` to queries.getParams().
// E.g.
// queries.getParams('paragraph--alert_single'))
//
// We don't need this here:
// export const params: QueryParams<null> = () => {
//   return queries
//     .getParams()
//     .addInclude([
//       'field_alert_block_reference',
//       'field_alert_block_reference.field_alert_content',
//       'field_alert_non_reusable_ref',
//       'field_alert_non_reusable_ref.field_va_paragraphs',
//     ])
// }
//
// ...but it seems like we might want to define
// paragraph-specific `includes` within the paragraph's
// query file (this file), and then somehow reference these in
// the node query when we need to include fields of a specific
// paragraph type.
//
// E.g.
// export const include = [
//   'field_alert_block_reference',
//   'field_alert_block_reference.field_alert_content',
//   'field_alert_non_reusable_ref',
//   'field_alert_non_reusable_ref.field_va_paragraphs',
// ]
//
// /* Some util file*/
// export const getIncludedSubFields = (
//   parentField: string,
//   subFields: string[]
// ) => [parentField, ...subFields.map((subField) => `${parentField}.${subField}`)]
//
// /* src/data/queries/resourcesSupport.ts */
// import { include } from '@/data/queries/alertSingle'
// export const params: QueryParams<null> = () => {
//   return queries.getParams().addInclude([
//     ...getIncludedSubFields('field_alert_single', include)
//   ])
// }

export const formatter: QueryFormatter<ParagraphAlertSingle, AlertSingle> = (
  entity: ParagraphAlertSingle
) => {
  return {
    type: entity.type as AlertSingle['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    alertSelection:
      entity.field_alert_selection as AlertSingle['alertSelection'],
    blockReference: queries.formatData(
      'block--alert',
      entity.field_alert_block_reference
    ) as AlertBlock,
    nonReusableRef: formatParagraph(
      entity.field_alert_non_reusable_ref
    ) as AlertNonReusable,
  }
}
