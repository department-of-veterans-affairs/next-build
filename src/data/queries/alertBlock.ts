// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter } from 'next-drupal-query'
import { AlertBlock, AlertType } from '@/types/formatted/alert'

export const formatter: QueryFormatter<BlockAlert, AlertBlock> = (
  entity: BlockAlert
) => {
  if (!entity) {
    return null
  }

  return {
    alertType: (entity.field_alert_type === 'information'
      ? 'info'
      : entity.field_alert_type) as AlertType,
    id: entity.id,
    title: entity.field_alert_title,
    content: {
      header:
        entity.field_alert_content.field_text_expander ??
        entity.field_alert_heading,
      text: entity.field_alert_content?.field_wysiwyg?.processed ?? '',
    },
  }
}
