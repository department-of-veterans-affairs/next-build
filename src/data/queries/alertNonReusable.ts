// Define the query params for fetching block--alert.
import { BlockAlert } from '@/types/drupal/block'
import { QueryFormatter } from 'next-drupal-query'
import { AlertType, AlertNonReusable } from '@/types/formatted/alert'
import { ParagraphNonReusableAlert } from '@/types/drupal/paragraph'
import { formatParagraph } from '@/lib/drupal/paragraphs'

export const formatter: QueryFormatter<
  ParagraphNonReusableAlert,
  AlertNonReusable
> = (entity: ParagraphNonReusableAlert) => {
  if (!entity) {
    return null
  }

  return {
    type: entity.type as AlertNonReusable['type'],
    id: entity.id,
    entityId: entity.drupal_internal__id,
    alertType: (entity.field_alert_type === 'information'
      ? 'info'
      : entity.field_alert_type) as AlertType,
    heading: entity.field_alert_heading,
    paragraphs: entity.field_va_paragraphs.map(formatParagraph),
  }
}
