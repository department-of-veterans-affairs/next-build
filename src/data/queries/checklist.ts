import { QueryData, QueryFormatter, QueryParams } from 'next-drupal-query'
import { DrupalJsonApiParams } from 'drupal-jsonapi-params'
import { queries } from '.'
import { formatParagraph } from '@/lib/drupal/paragraphs'
import { PARAGRAPH_RESOURCE_TYPES, RESOURCE_TYPES } from '@/lib/constants/resourceTypes'
import { ExpandedStaticPropsContext } from '@/lib/drupal/staticProps'
import {
  entityBaseFields,
  fetchSingleEntityOrPreview,
} from '@/lib/drupal/query'
import { getNestedIncludes } from '@/lib/utils/queries'

// Types
import { AlertSingle } from '@/types/formatted/alert'
import { Button } from '@/types/formatted/button'
import { Checklist } from '@/types/formatted/checklist'
import { ChecklistItem } from '@/types/formatted/checklistItem'
import { ContactInfo } from '@/types/formatted/contactInfo'
import { NodeChecklist } from '@/types/drupal/node'
import { TaxonomyTermLcCategories } from '@/types/drupal/taxonomy_term'

export const params: QueryParams<null> = () => {
  return new DrupalJsonApiParams().addInclude([
    ...getNestedIncludes(
      'field_alert_single',
      PARAGRAPH_RESOURCE_TYPES.ALERT_SINGLE
    ),
    'field_buttons',
    'field_checklist.field_checklist_sections',
    ...getNestedIncludes(
      'field_contact_information',
      PARAGRAPH_RESOURCE_TYPES.CONTACT_INFORMATION
    ),
    'field_related_benefit_hubs',
    'field_related_information',
    ...getNestedIncludes(
      'field_tags',
      PARAGRAPH_RESOURCE_TYPES.AUDIENCE_TOPICS
    ),
  ])
}

export type ChecklistDataOpts = {
  id: string
  context?: ExpandedStaticPropsContext
}

export const data: QueryData<ChecklistDataOpts, NodeChecklist> = async (
  opts
): Promise<NodeChecklist> => {
const entity = (await fetchSingleEntityOrPreview(
    opts,
    RESOURCE_TYPES.CHECKLIST,
    params
  )) as NodeChecklist

  return entity
}

export const formatter: QueryFormatter<NodeChecklist, Checklist> = (
  entity: NodeChecklist
) => {
  return {
    ...entityBaseFields(entity),
    alert: formatParagraph(entity.field_alert_single) as AlertSingle,
    buttons: entity.field_buttons.map?.(formatParagraph) as Button[],
    checklist: entity.field_checklist.field_checklist_sections?.map(section =>
      queries.formatData(
        PARAGRAPH_RESOURCE_TYPES.CHECKLIST,
        section
      )
    ) as ChecklistItem[],
    contactInformation: formatParagraph(
      entity.field_contact_information
    ) as ContactInfo,
    intro: entity.field_intro_text_limited_html.processed,
    otherCategories: entity.field_other_categories.map?.(formatParagraph) as TaxonomyTermLcCategories[],
    primaryCategory: entity.field_primary_category,
    repeatButtons: entity.field_buttons_repeat,
  }
}
