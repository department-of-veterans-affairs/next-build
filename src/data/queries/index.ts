import { createQueries } from 'next-drupal-query'
import * as NewsStory from './newsStory'
import * as NewsStoryTeaser from './newsStoryTeaser'
import * as StoryListing from './storyListing'
import * as QuestionAnswer from './questionAnswer'
import * as ExpandableText from './expandableText'
import * as LinkTeaser from './linkTeaser'
import * as MediaImage from './mediaImage'
import * as Banner from './banners'
import * as PersonProfile from './personProfile'
import * as Button from './button'
import * as AudienceTopics from './audienceTopics'
import * as Alert from './alert'
import * as EmailContact from './emailContact'
import * as BenefitsHub from './benefitsHub'
import * as Wysiwyg from './wysiwyg'
import * as StaticPathResources from './staticPathResources'
import * as HeaderFooter from './headerFooter'

export const queries = createQueries({
  // standard Drupal entity data queries
  'node--news_story': NewsStory,
  'node--news_story--teaser': NewsStoryTeaser,
  'node--story_listing': StoryListing,
  'node--q_a': QuestionAnswer,
  'node--person_profile': PersonProfile,
  'node--landing_page': BenefitsHub, // "Benefits Hub Landing Page"
  'paragraph--audience_topics': AudienceTopics,
  'paragraph--button': Button,
  'paragraph--email_contact': EmailContact,
  'paragraph--expandable_text': ExpandableText,
  'paragraph--link_teaser': LinkTeaser,
  'paragraph--rich_text_char_limit_1000': Wysiwyg,
  'paragraph--wysiwyg': Wysiwyg,
  'media--image': MediaImage,
  'block--alert': Alert,

  // Custom queries 
  'banner--alerts_lookup': Banner,
  'header-footer-data': HeaderFooter,

  // Static Path Generation
  'static-path-resources': StaticPathResources,
})
