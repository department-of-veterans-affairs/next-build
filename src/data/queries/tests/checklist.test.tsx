/**
 * @jest-environment node
 */

import { Checklist } from '@/types/drupal/node'
import { queries } from '@/data/queries'
import mockData from '@/mocks/checklist.mock.json'
import { params } from '../checklist';

const ChecklistMock: Checklist = mockData

describe('DrupalJsonApiParams configuration', () => {
  test('params function sets the correct include fields', () => {
    const paramsInstance = params()
    const queryString = decodeURIComponent(paramsInstance.getQueryString())

    const expectedIncludes = [
      'field_alert_single',
      'field_buttons',
      'field_checklist.field_checklist_sections',
      'field_contact_information',
      'field_related_benefit_hubs',
      'field_related_information',
      'field_tags'
    ]

    expectedIncludes.forEach((include) => {
      expect(queryString).toMatch(new RegExp(`include=.*${include}`))
    })
  })
})

describe('Checklist formatData', () => {
  test('outputs formatted data', () => {
    expect(
      queries.formatData(
        'node--checklist',
        ChecklistMock
      )
    ).toMatchSnapshot()
  })
})
