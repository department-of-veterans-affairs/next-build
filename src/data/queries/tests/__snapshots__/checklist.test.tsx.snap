// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checklist formatData outputs formatted data 1`] = `
{
  "alert": null,
  "benefitsHubLinks": [
    {
      "id": "fd822f25-788f-489c-a0fc-b1f7ad6c20d5",
      "label": undefined,
      "path": undefined,
      "teaserText": undefined,
      "title": undefined,
    },
    {
      "id": "ea251553-7aa5-4613-8824-47965c489a86",
      "label": undefined,
      "path": undefined,
      "teaserText": undefined,
      "title": undefined,
    },
    {
      "id": "18ca8756-9f78-43ef-95ea-04bd106063c5",
      "label": undefined,
      "path": undefined,
      "teaserText": undefined,
      "title": undefined,
    },
  ],
  "breadcrumbs": [
    {
      "options": [],
      "title": "Home",
      "uri": "https://cms-p97uyeo6apm3pwstjzyqxw0fh0wbjsc1.demo.cms.va.gov/",
    },
    {
      "options": [],
      "title": "Resources and Support",
      "uri": "https://cms-p97uyeo6apm3pwstjzyqxw0fh0wbjsc1.demo.cms.va.gov/resources",
    },
    {
      "options": [],
      "title": "Test Checklist Page",
      "uri": "https://cms-p97uyeo6apm3pwstjzyqxw0fh0wbjsc1.demo.cms.va.gov/resources/test-checklist-page",
    },
  ],
  "buttons": [
    {
      "entityId": undefined,
      "id": "818d2959-a00f-4aec-9366-5164eeeae67a",
      "label": null,
      "type": "paragraph--button",
      "url": null,
    },
    {
      "entityId": undefined,
      "id": "5cb4f676-0e67-4929-81c4-c5fac1b8ee2e",
      "label": null,
      "type": "paragraph--button",
      "url": null,
    },
  ],
  "checklist": undefined,
  "contactInformation": null,
  "entityId": 73664,
  "entityPath": "/resources/test-checklist-page",
  "id": "1f681975-6c2a-4d42-a294-785e165c7eff",
  "intro": "<p>This is the page introduction. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce cursus malesuada neque, id mattis ligula venenatis dapibus. Nulla purus ligula, ultricies id maximus eu, venenatis eu nulla. Suspendisse auctor turpis non iaculis molestie. Mauris luctus libero erat, sed tempor est hendrerit eu. Cras sodales egestas mi, et porta sapien pharetra eget. Pellentesque ornare metus sed eros ornare porta. Sed egestas, felis sed imperdiet commodo, eros mauris venenatis velit, vitae mattis augue velit sed lectus.</p>
",
  "lastUpdated": "2024-10-16T20:04:12+00:00",
  "metatags": [
    {
      "attributes": {
        "content": "Test Checklist Page | Veterans Affairs",
        "name": "title",
      },
      "tag": "meta",
    },
    {
      "attributes": {
        "content": "This is the page introduction. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce cursus malesuada neque, id mattis ligula venenatis dapibus. Nulla purus ligula, ultricies id maximus eu, venenatis eu nulla. Suspendisse auctor turpis non iaculis molestie. Mauris luctus libero erat, sed tempor est hendrerit eu. Cras sodales egestas mi, et porta sapien pharetra eget. Pellentesque ornare metus sed eros ornare porta. Sed egestas, felis sed imperdiet commodo, eros mauris venenatis velit, vitae mattis augue velit sed lectus.",
        "name": "description",
      },
      "tag": "meta",
    },
    {
      "attributes": {
        "href": "https://www.va.gov/img/design/logo/va-og-image.png",
        "rel": "image_src",
      },
      "tag": "link",
    },
    {
      "attributes": {
        "content": "Veterans Affairs",
        "property": "og:site_name",
      },
      "tag": "meta",
    },
    {
      "attributes": {
        "content": "Test Checklist Page | Veterans Affairs",
        "property": "og:title",
      },
      "tag": "meta",
    },
    {
      "attributes": {
        "content": "This is the page introduction. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce cursus malesuada neque, id mattis ligula venenatis dapibus. Nulla purus ligula, ultricies id maximus eu, venenatis eu nulla. Suspendisse auctor turpis non iaculis molestie. Mauris luctus libero erat, sed tempor est hendrerit eu. Cras sodales egestas mi, et porta sapien pharetra eget. Pellentesque ornare metus sed eros ornare porta. Sed egestas, felis sed imperdiet commodo, eros mauris venenatis velit, vitae mattis augue velit sed lectus.",
        "property": "og:description",
      },
      "tag": "meta",
    },
    {
      "attributes": {
        "content": "https://www.va.gov/img/design/logo/va-og-image.png",
        "property": "og:image",
      },
      "tag": "meta",
    },
    {
      "attributes": {
        "content": "U.S. Department of Veterans Affairs",
        "property": "og:image:alt",
      },
      "tag": "meta",
    },
    {
      "attributes": {
        "content": "summary_large_image",
        "name": "twitter:card",
      },
      "tag": "meta",
    },
    {
      "attributes": {
        "content": "@DeptVetAffairs",
        "name": "twitter:site",
      },
      "tag": "meta",
    },
    {
      "attributes": {
        "content": "Test Checklist Page | Veterans Affairs",
        "name": "twitter:title",
      },
      "tag": "meta",
    },
    {
      "attributes": {
        "content": "https://www.va.gov/img/design/logo/va-og-image.png",
        "name": "twitter:image",
      },
      "tag": "meta",
    },
    {
      "attributes": {
        "content": "U.S. Department of Veterans Affairs",
        "name": "twitter:image:alt",
      },
      "tag": "meta",
    },
  ],
  "moderationState": "published",
  "published": true,
  "relatedInformation": undefined,
  "repeatButtons": true,
  "tags": null,
  "title": "Test Checklist Page",
  "type": "node--checklist",
}
`;
